lib/ecto/multi.ex

```diff
@@ -0,0 +1,121 @@
+defmodule Ecto.Multi do
+  alias __MODULE__
+  alias Ecto.Changeset
+
+  defstruct operations: [], names: MapSet.new
+
+  @type run :: (t, Keyword.t -> {:ok | :error, any})
+  @type operation :: Changeset.t | run
+  @type name :: atom
+  @type t :: %__MODULE__{operations: [{name, operation}]}
+
+  @spec new :: t
+  def new do
+    %Multi{}
+  end
+
+  @spec insert(t, name, Changeset.t | Model.t) :: t
+  def insert(multi, name, %Changeset{} = changeset) do
+    add_changeset(multi, :insert, name, changeset)
+  end
+
+  def insert(multi, name, struct) do
+    add_changeset(multi, :insert, name, Changeset.change(struct))
+  end
+
+  @spec update(t, name, Changeset.t) :: t
+  def update(multi, name, %Changeset{} = changeset) do
+    add_changeset(multi, :update, name, changeset)
+  end
+
+  @spec delete(t, name, Changeset.t | Model.t) :: t
+  def delete(multi, name, %Changeset{} = changeset) do
+    add_changeset(multi, :delete, name, changeset)
+  end
+
+  def delete(multi, name, struct) do
+    add_changeset(multi, :delete, name, Changeset.change(struct))
+  end
+
+  defp add_changeset(multi, action, name, changeset) do
+    add_operation(multi, name, put_action(changeset, action))
+  end
+
+  defp put_action(%{action: nil} = changeset, action) do
+    %{changeset | action: action}
+  end
+
+  defp put_action(%{action: action} = changeset, action) do
+    changeset
+  end
+
+  defp put_action(%{action: original}, action) do
+    raise ArgumentError, "you provided a changeset with an action already set " <>
+      "to #{original} when trying to #{action} it"
+  end
+
+  @spec run(t, name, run) :: t
+  def run(multi, name, run) when is_function(run, 2) do
+    add_operation(multi, name, run)
+  end
+
+  @spec run(t, name, module, function, args) :: t
+    when function: atom, args: [any]
+  def run(multi, name, mod, fun, args) do
+    add_operation(multi, name, {mod, fun, args})
+  end
+
+  defp add_operation(%Multi{operations: operations, names: names} = multi, name, operation)
+      when is_atom(name) do
+    if MapSet.member?(names, name) do
+      raise "#{name} is already a member of the Ecto.Multi: \n#{inspect multi}"
+    else
+      %{multi | operations: [{name, operation} | operations],
+                names: MapSet.put(names, name)}
+    end
+  end
+
+  @spec apply(t, Ecto.Repo.t, wrap, return, Keyword.t) ::
+      {:ok, results} | {:error, {name, errors}}
```

Comment:
We can't really with rollback. We would need to repack the error after returning from transaction.